src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -526,76 +523,90 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
-						const promiseResult = await promise;
-						if (promiseResult.exitCode === 0) {
-							promise = Promise.resolve(promiseResult);
+		const mapKey = task.getMapKey();
+
+		// It's important that we add this task's entry to _activeTasks before
+		// any of the code in the then runs (see #180541 and #180578). Wrapping
+		// it in Promise.resolve().then() ensures that.
+		const promise = Promise.resolve().then(async () => {
+			alreadyResolved = alreadyResolved ?? new Map<string, string>();
+			const promises: Promise<ITaskSummary>[] = [];
+			if (task.configurationProperties.dependsOn) {
+				liveDependencies = new Set(liveDependencies).add(task.getCommonTaskId());
+				for (const dependency of task.configurationProperties.dependsOn) {
+					const dependencyTask = await resolver.resolve(dependency.uri, dependency.task!);
+					if (dependencyTask) {
+						this._adoptConfigurationForDependencyTask(dependencyTask, task);
+						let promise;
+						const commonKey = dependencyTask.getCommonTaskId();
+						if (liveDependencies.has(commonKey)) {
+							this._showDependencyCycleMessage(dependencyTask);
+							promise = Promise.resolve<ITaskSummary>({});
 						} else {
-							promise = Promise.reject(promiseResult);
-							break;
+							promise = encounteredTasks.get(commonKey);
+							if (!promise) {
+								const key = dependencyTask.getMapKey();
+								promise = this._activeTasks[key] ? this._getDependencyPromise(this._activeTasks[key]) : undefined;
+							}
+						}
+						if (!promise) {
+							this._fireTaskEvent(TaskEvent.create(TaskEventKind.DependsOnStarted, task));
+							promise = this._executeDependencyTask(dependencyTask, resolver, trigger, liveDependencies, encounteredTasks, alreadyResolved);
+						}
+						encounteredTasks.set(commonKey, promise);
+						promises.push(promise);
+						if (task.configurationProperties.dependsOrder === DependsOrder.sequence) {
+							const promiseResult = await promise;
+							if (promiseResult.exitCode !== 0) {
+								break;
+							}
 						}
+					} else {
+						this._log(nls.localize('dependencyFailed',
+							'Couldn\'t resolve dependent task \'{0}\' in workspace folder \'{1}\'',
+							Types.isString(dependency.task) ? dependency.task : JSON.stringify(dependency.task, undefined, 0),
+							dependency.uri.toString()
+						));
+						this._showOutput();
 					}
-					promises.push(promise);
-				} else {
-					this._log(nls.localize('dependencyFailed',
-						'Couldn\'t resolve dependent task \'{0}\' in workspace folder \'{1}\'',
-						Types.isString(dependency.task) ? dependency.task : JSON.stringify(dependency.task, undefined, 0),
-						dependency.uri.toString()
-					));
-					this._showOutput();
 				}
 			}
-		}
 
-		if ((ContributedTask.is(task) || CustomTask.is(task)) && (task.command)) {
-			return Promise.all(promises).then((summaries): Promise<ITaskSummary> | ITaskSummary => {
-				encounteredDependencies.delete(task.getCommonTaskId());
```

Comment:
Where are task IDs removed from `liveDependencies` now?

Reply:
I already commented on the "fix", but to clarify, they don't need to be removed, because the passed in set is never modified. The passed in set is cloned, and the new entry is added to the clone, and the clone is passed down to the dependent tasks. Once the dependent tasks have all been launched, the clone is dead, and will be garbage collected.

Note that due to the asynchronous nature of this, there's no way to maintain a single set that detects cycles in the directed graph (which are errors), while not falsely identifying cycles in the undirected graph (which are fine - eg A depends on B, C which each depend on D).